### PR TITLE
Changes to script create_runtime_policy.sh, fixes #1426

### LINKED ISSUE #1426: 2 outputs built by the script create_runtime_policy.sh instead only one
```
[root@fedora project]# ./keylime/scripts/create_runtime_policy.sh -o runtime_policy_default.json -a sha256sum
==> jeu. 13 juil. 2023 10:28:26 CEST - ./keylime/scripts/create_runtime_policy.sh - Writing allowlist to runtime_policy_default.json with sha256sum...
==> jeu. 13 juil. 2023 10:28:26 CEST - ./keylime/scripts/create_runtime_policy.sh - Adding boot agregate from /sys/kernel/security/ima/ascii_runtime_measurements on allowlist runtime_policy_default.json...
==> jeu. 13 juil. 2023 10:28:26 CEST - ./keylime/scripts/create_runtime_policy.sh - Adding all appropriate files from / on allowlist runtime_policy_default.json...
21
==> jeu. 13 juil. 2023 10:39:13 CEST - ./keylime/scripts/create_runtime_policy.sh - Creating allowlist for init ram disks found under "/boot/"...
==> jeu. 13 juil. 2023 10:39:13 CEST - ./keylime/scripts/create_runtime_policy.sh - extracting /boot//initramfs-0-rescue-4a8280baba9d45a99c4014c8c6911545.img
==> jeu. 13 juil. 2023 10:39:19 CEST - ./keylime/scripts/create_runtime_policy.sh - extracting /boot//initramfs-6.2.9-300.fc38.x86_64.img
==> jeu. 13 juil. 2023 10:39:22 CEST - ./keylime/scripts/create_runtime_policy.sh - extracting /boot//initramfs-6.3.12-200.fc38.x86_64.img
==> jeu. 13 juil. 2023 10:39:24 CEST - ./keylime/scripts/create_runtime_policy.sh - Converting created allowlist to Keylime runtime policy
INFO:keylime.config:Reading configuration from ['/etc/keylime/logging.conf']
runtime_policy_default.json_allowlist is not JSON-formatted; attempting to convert to IMA policy from flat file format
[root@fedora project]# ll
total 24
drwxr-xr-x. 1 xxx xxx   992 12 juil. 14:53 keylime
-rw-r--r--. 1 root     root        81 13 juil. 10:28 runtime_policy_default.json_allowlist
drwxr-xr-x. 1 xxx xxx   594 13 juil. 10:19 rust-keylime
[root@fedora project]# ll /runtime_policy_default.json_allowlist
-rw-r--r--. 1 root root 28045700 13 juil. 10:39 /runtime_policy_default.json_allowlist
```
file : ~/project/runtime_policy_default.json_allowlist
```
58799d0a6dfcb7e3a911000ba587e549ef128642895890e6fd0f3e61d3f8cb4d  boot_aggregate
```

file : /runtime_policy_default.json_allowlist
```
c445d13771b6fe32f728350269967bc4938e8b395fcc57bca0358ee0c2e4bcf0  //boot/initramfs-6.3.12-200.fc38.x86_64.img
380c85ce1cadbac65fe62faa8d3c4caf4dd75f82ccd267b187715a6316d12a6f  //boot/symvers-6.3.12-200.fc38.x86_64.gz
9470c9b73d165b203726652fb30996676aebf6ed29e6083e4e39d5c2c83547ba  //boot/config-6.3.12-200.fc38.x86_64
744218da9e6c1302c9d436070c1094a829ad9b5821cc714a5a2835877e657327  //boot/vmlinuz-0-rescue-4a8280baba9d45a99c4014c8c6911545
...
```
In add, I understand that the final file should be ~/project/runtime_policy_default.json according to the extract from the script 'create_runtime_policy.sh'

``` 
# Convert to runtime policy
announce "Converting created allowlist to Keylime runtime policy"
CONVERT_CMD_OPTS="--allowlist ${OUTPUT}_allowlist --output_file $OUTPUT"

[ -f $EXCLUDE_LIST ] && CONVERT_CMD_OPTS="$CONVERT_CMD_OPTS --excludelist $EXCLUDE_LIST"

python3 $WORKING_DIR/../keylime/cmd/convert_runtime_policy.py $CONVERT_CMD_OPTS
```
The final file is not found.
```
[xxx@fedora project]$ sudo find / | grep runtime_policy_default.json
[sudo] Mot de passe de xxx : 
/home/xxx/project/runtime_policy_default.json_allowlist
find: ‘/run/user/1000/doc’: Permission non accordée
find: ‘/run/user/1000/gvfs’: Permission non accordée
/runtime_policy_default.json_allowlist

```

**suggested fix for the 1st problem :**
```
create_runtime_policy.sh: 187 : popd > /dev/null 2>&1
create_runtime_policy.sh: ... : cat $ROOTFS_LOC/${OUTPUT}_allowlist >> ${OUTPUT}_allowlist
create_runtime_policy.sh: ... : rm -rf $ROOTFS_LOC/${OUTPUT}_allowlist
```
**workaround for the 2nd problem :**
```
# python3 ./keylime/keylime/cmd/convert_runtime_policy.py --allowlist runtime_policy_default.json_allowlist --output_file runtime_policy_default.json
# python3 ./keylime/keylime/cmd/convert_runtime_policy.py --allowlist runtime_policy_default.json_allowlist --output_file runtime_policy_exclude.json  --excludelist excludelist
[root@fedora project]# ll
total 85032
-rw-r--r--. 1 root     root            6 13 juil. 14:08 excludelist
drwxr-xr-x. 1 xxx xxx      992 12 juil. 14:53 keylime
-rw-r--r--. 1 root     root        14164 13 juil. 09:30 measured_boot_reference_state.json
-rw-r--r--. 1 root     root            3 13 juil. 09:28 measured_boot_reference_state.txt
-rw-r--r--. 1 root     root            0 13 juil. 14:20 payload
-rw-r--r--. 1 root     root     29498523 13 juil. 14:10 runtime_policy_default.json
-rw-r--r--. 1 root     root     28045770 13 juil. 14:28 runtime_policy_default.json_allowlist
-rw-r--r--. 1 root     root     29498508 13 juil. 14:28 runtime_policy_exclude.json
drwxr-xr-x. 1 xxx xxx      594 13 juil. 10:19 rust-keylime
```

Also, some hashes always have the '\' prefix and will create the following error with keylime_tenant.

```
"Runtime policy is malformatted: '\\\\7eb1547eeae0887bd58680e0c41a6f7bfe47babd356a98b91b4d39a5ade31501' does not match '^[0-9a-f]{40,128}$'"
```

```
[root@fedora project]# grep '^\\' runtime_policy_default.json_allowlist
\7eb1547eeae0887bd58680e0c41a6f7bfe47babd356a98b91b4d39a5ade31501  //etc/systemd/system/dev-virtio\\x2dports-org.qemu.guest_agent.0.device.wants/qemu-guest-agent.service
\4869b965f6f1cdefc60c19b48430486b8fdcc77b14201b0588fc37306d455967  //usr/lib/systemd/system/run-vmblock\\x2dfuse.mount
\2136146f9e0019eb33c68c84b7db3160f9a430fd9b497b58a5396e24226a2417  //usr/lib/systemd/system/system-systemd\\x2dcryptsetup.slice
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-chromium\\x2dbrowser-.scope.d/99-cgroupify.conf
\0b4f552c2983c419732369a8d561ed470e6f41c455c4dfa00315de446bf60ebf  //usr/lib/systemd/user/app-chromium\\x2dbrowser@@.service.d/99-cgroupify.conf
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-glib-chromium\\x2dbrowser-.scope.d/99-cgroupify.conf
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-glib-google\\x2dchrome-.scope.d/99-cgroupify.conf
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-gnome-chromium\\x2dbrowser-.scope.d/99-cgroupify.conf
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-gnome-google\\x2dchrome-.scope.d/99-cgroupify.conf
\7c1faefddbf9aa76721c4fe44ee3fbde9fff598cbbcb6cd8b21a1b62f0ad39d9  //usr/lib/systemd/user/app-google\\x2dchrome-.scope.d/99-cgroupify.conf
\0b4f552c2983c419732369a8d561ed470e6f41c455c4dfa00315de446bf60ebf  //usr/lib/systemd/user/app-google\\x2dchrome@@.service.d/99-cgroupify.conf
[root@fedora project]# python3 ./keylime/keylime/cmd/convert_runtime_policy.py -a runtime_policy_default.json_allowlist -o runtime_policy_default.json_allowlist 
INFO:keylime.config:Reading configuration from ['/etc/keylime/logging.conf']
runtime_policy_default.json_allowlist is not JSON-formatted; attempting to convert to IMA policy from flat file format
```

**suggested fix for the 3rd problem :**
```
[root@fedora project]# sed -i "s/^\\\//g" runtime_policy_default.json_allowlist
[root@fedora project]# grep '^\\' runtime_policy_default.json_allowlist
[root@fedora project]# 
```

### PR DESCRIPTION
**Pull request by maugustosilva**
In addition to that, automatically detects the algorithm used for boot aggregate (if one is specified) and exits with error if this differs from the one used with the -a/--algo option

### CODE REVIEW

File: scripts/create_runtime_policy.sh [hunk 1]
```diff
@@ -4,6 +4,30 @@
...
+    pushd `dirname "$0"` > /dev/null 2>&1
+fi
+KCRP_BASE_DIR=$(pwd)
+if [ $0 != "-bash" ] ; then
+    popd 2>&1 > /dev/null
+fi
+KCRP_BASE_DIR=$KCRP_BASE_DIR/..
+
+function detect_hash {
+    hashstr=$1
```
Comment (stefanberger) [inline scripts/create_runtime_policy.sh:18]: local hashstr=$1

File: scripts/create_runtime_policy.sh [hunk 2]
```diff
@@ -23,14 +47,14 @@ function valid_algo {
 INITRAMFS_TOOLS_GIT=https://salsa.debian.org/kernel-team/initramfs-tools.git
 INITRAMFS_TOOLS_VER="master"
 
-WORKING_DIR=$(readlink -f "$0")
-WORKING_DIR=$(dirname "$WORKING_DIR")
-
 # All defaults
 ALGO=sha1sum
+WORK_DIR=/tmp/kcrp
```
Comment (stefanberger) [inline scripts/create_runtime_policy.sh:52]: Use `WORK_DIR=$(mktemp -d)` to create a temporary directory ? And then do
```
function cleanup() {
  if -d ${WOKRDIR}; then
    rm -rf ${WORKDIR}
  fi
}
trap cleanup EXIT
```

File: scripts/create_runtime_policy.sh [hunk 3]
```diff
@@ -154,40 +178,47 @@ then
     exit 1
 fi
 
-rm -rf ${OUTPUT}_allowlist
-rm -rf ${OUTPUT}
+rm -rf $ALLOWLIST_DIR
+rm -rf $INITRAMFS_STAGING_DIR
+rm -rf $OUTPUT_DIR
```
Comment (stefanberger) [inline scripts/create_runtime_policy.sh:183]: with a temp workdir all of these `rm -rf` could go.

File: scripts/create_runtime_policy.sh [hunk 4]
```diff
@@ -221,25 +252,30 @@ then
...
-        find -type f -exec $ALGO "./{}" \; | sed "s| \./\./| /|" >> ${OUTPUT}_allowlist
+        find -type f -exec $ALGO "./{}" \; | sed "s| \./\./| /|" >> $ALLOWLIST_DIR/${OUTPUT}
     done
 fi
 
 # A bit of cleanup on the resulting file (among other problems, sha256sum might output a hash with the prefix '//')
-sed -i "s^ //^ /^g"  ${OUTPUT}_allowlist 
-sed -i 's^"\\\\^"^g' ${OUTPUT}_allowlist
+sed -i "s^ //^ /^g"  $ALLOWLIST_DIR/${OUTPUT} 
+sed -i "s/^\\\//g" $ALLOWLIST_DIR/${OUTPUT}
```
Comment (stefanberger) [inline scripts/create_runtime_policy.sh:265]: So this is removing a leading `\/` ? Mabe add to documentation above.

Comment (maugustosilva) [reply scripts/create_runtime_policy.sh:266]: Just added a better explanation to both

**Review by aplanas — APPROVED** [review]

**Pull request merged**